docs/proposals/20190704-AutomatedUpgrades.md

```diff
@@ -0,0 +1,206 @@
+This is part of the solution to remove the need for `ROOT-A` and `ROOT-B` partitions.
+It also allows us to dedicate the owned block device to data that can be reproduced (e.g. containerd data, kubelet data, etcd data, CNI data, etc.).
+
+### Using `squashfs` for the `rootfs`
+
+For performance reasons, and a smaller footprint, we can create our `rootfs` as a `squashfs` image.
+The way in which we can run the `rootfs` is by mounting it using a loopback device.
+Once mounted, we have two options:
+
+- perform a `switch_root` into the mounted `rootfs`
+- `fork` a `machined` process without performing a `switch_root`
+
+Another benefit of `squashfs` is that it allows us to retain our read only `rootfs` without having to mount it as such.
+
+> TODO: More research needs to be done here.
+> The driving force behind this decision is figuring out how we can return control back to `init` in order to reset a node.
+
+### Slimming down `rootfs`
+
+#### Proposal 1
+
+Today our `rootfs` contains large tarballs from Kubernetes.
+In this implementation we will remove these tarballs to a dedicated `cache` tarball.
+This allows us to considerably reduce the size of the `rootfs`.
+
+The downside to this is that the tarballs are no longer on a read only partition and can be overwritten, taking us a step away from our immutability goals.
+
+#### Proposal 2
+
+An alternative approach is to use compression.
+We currently do not compress the container tarballs.
+The idea here is that by compressing the container images, and then creating a `squashfs`, we can reduce the `rootfs` down to an acceptable size.
+This has the advantage that everything is nicely bundled together.
+
+A downside is that it will take the kernel longer to decompress the `initramfs`.
+Another downside is that we would require at least 1GB of RAM to run talos.
+
+#### Proposal 3
+
+Finally, we could also remove the Kubernetes tarballs entirely and let the `kubelet` download them.
+This also is a shift from our immutability goals, but in this case we can embrace configurability at the Kubernetes layer.
+The disadvantage to this is that this introduces a number a variables in the number of things that can go wrong.
+Having "approved" tarballs packaged with Talos allows us to run conformance tests against exactly what users will use.
+
+### A New Partition Scheme
+
+With the move to run the `rootfs` in memory, we will remove the need for `ROOT-A` and `ROOT-B` partitions.
+In the case of cloud installs, or where users may want to boot without PXE/iPXE, we will require a boot partition.
+In all cases we require a data partition.
+
+I also propose we rename the `DATA` partition to `EPHEMERAL` to make it clear that this partition should not be used to store application data.
+Our suggestion to users who do want persistent storage on the host is to mount another disk and to use that.
+Talos will make the guarantee to only manage the "owned" block device.
+
+### Performing Upgrades in a Container
```

Comment:
An alternative we discussed in a community meeting is to _not_ run upgrades in a container, but to still used a container image to deliver the artifacts. This will remove the need for `containerd` in the early boot environment.